jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +264,265 @@ public InvocationType getInvocationType()
             {
                 return callback.getInvocationType();
             }
-        });
+        };
+        _contentDemander.onContent(content);
+        boolean handle = onContent(content);
 
         boolean endStream = frame.isEndStream();
         if (endStream)
         {
             boolean handleContent = onContentComplete();
+            // this will generate EOF -> must happen before wakeup
             boolean handleRequest = onRequestComplete();
             handle |= handleContent | handleRequest;
         }
 
+        boolean woken = getRequest().getHttpInput().wakeup();
```

Comment:
I think this is wrong because it means a spurious wakeup can occur even if `needContent` has not been called.

A call to `produceContent` can do demand, so `onData` may be called at any time afterwards... so it is wrong for it to wakeup HttpInput if needContent has not been called.

This class needs to know if `needContent` has been called and only do a `wakeup` if it has.

Reply:
The more I think about this, the more I'm convinced we need a dedicated `onContentAvailable` callback that is only called if `needContent` has been called.

Reply:
We could almost pass a `Callback` in needContent to make the contract clear!